### CapStone Project Week 4
## Introduction

Moving to a new city is one of the most stressfull events in someone's life. As you leave your home, you don't know what you'll find on the other side. Where will you shop your grosseries, which school your kids will go to, will you find new friends? How long will it take before you feel at home again?

In this project we'll suggest where someone living in Paris should move in Brussels.



## Data

We'll use the postal codes of both cities to identify their neighborhoods. 
The list of postal codes of Paris may be found here: https://www.laposte.fr/particulier/outils/trouver-un-code-postal
Those of Brussels can be found here: https://www.bpost.be/site/fr/envoyer/adressage/rechercher-un-code-postal

Let's import them and see what they look like:

In [1]:
# The code was removed by Watson Studio for sharing.

Code Postal  Ville
0         75001  PARIS
1         75002  PARIS
2         75003  PARIS
3         75004  PARIS
4         75005  PARIS
5         75006  PARIS
6         75007  PARIS
7         75008  PARIS
8         75009  PARIS
9         75010  PARIS
10        75011  PARIS
11        75012  PARIS
12        75013  PARIS
13        75014  PARIS
14        75015  PARIS
15        75016  PARIS
16        75017  PARIS
17        75018  PARIS
18        75019  PARIS
19        75020  PARIS
20        75116  PARIS

Postal Code ***75116*** actually corresponds to the same neighborhoods as ***75016*** so we'll drop it

In [2]:
paris_codes_df.drop([20], inplace = True)
paris_codes_df

Code Postal  Ville
0         75001  PARIS
1         75002  PARIS
2         75003  PARIS
3         75004  PARIS
4         75005  PARIS
5         75006  PARIS
6         75007  PARIS
7         75008  PARIS
8         75009  PARIS
9         75010  PARIS
10        75011  PARIS
11        75012  PARIS
12        75013  PARIS
13        75014  PARIS
14        75015  PARIS
15        75016  PARIS
16        75017  PARIS
17        75018  PARIS
18        75019  PARIS
19        75020  PARIS

In [14]:
paris_codes_df.rename(columns={"Code Postal": "Postal Code"}, inplace = True)
paris_codes_df.head()

Postal Code  Ville
0        75001  PARIS
1        75002  PARIS
2        75003  PARIS
3        75004  PARIS
4        75005  PARIS

The neighborhoods in Paris are known by their numbers. I'll add that information to the data and drop the column 'Ville' because it doesn't give much information.

In [17]:
paris_codes_df['Neighborhood'] = ["Paris 1", "Paris 2", "Paris 3", "Paris 4", "Paris 5",
                                 "Paris 6", "Paris 7", "Paris 8", "Paris 9", "Paris 10",
                                 "Paris 11", "Paris 12", "Paris 13", "Paris 14", "Paris 15",
                                 "Paris 16", "Paris 17", "Paris 18", "Paris 19", "Paris 20"]
paris_codes_df.drop("Ville", axis = 1, inplace=True)
paris_codes_df

Postal Code Neighborhood
0         75001      Paris 1
1         75002      Paris 2
2         75003      Paris 3
3         75004      Paris 4
4         75005      Paris 5
5         75006      Paris 6
6         75007      Paris 7
7         75008      Paris 8
8         75009      Paris 9
9         75010     Paris 10
10        75011     Paris 11
11        75012     Paris 12
12        75013     Paris 13
13        75014     Paris 14
14        75015     Paris 15
15        75016     Paris 16
16        75017     Paris 17
17        75018     Paris 18
18        75019     Paris 19
19        75020     Paris 20

In [3]:
print('There are {} neighborhoods in Paris'.format(paris_codes_df.shape[0]))

There are 20 neighborhoods in Paris


Let's see the postal codes information for the city of Brussels

In [40]:

body = client_75203cb60a4e41d4bb1debcd2578d37c.get_object(Bucket='machinelearning-donotdelete-pr-7gneoytsamib8b',Key='brussels_postal_codes.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

brussels_codes_df = pd.read_csv(body,';')
brussels_codes_df


Code postal               Localité Sous-commune     Commune principale  \
0          1000              Bruxelles          Non              BRUXELLES   
1          1020                 Laeken          Oui              BRUXELLES   
2          1030             Schaerbeek          Non             SCHAERBEEK   
3          1040              Etterbeek          Non              ETTERBEEK   
4          1050                Ixelles          Non                IXELLES   
5          1060           Saint-Gilles          Non           SAINT-GILLES   
6          1070             Anderlecht          Non             ANDERLECHT   
7          1080   Molenbeek-Saint-Jean          Non   MOLENBEEK-SAINT-JEAN   
8          1081             Koekelberg          Non             KOEKELBERG   
9          1082  Berchem-Sainte-Agathe          Non  BERCHEM-SAINTE-AGATHE   
10         1083              Ganshoren          Non              GANSHOREN   
11         1090                  Jette          Non                  JETTE   
12         1120    Neder-Over-Heembeek          Oui              BRUXELLES   
13         1130                  Haren          Oui              BRUXELLES   
14         1140                  Evere          Non                  EVERE   
15         1150    Woluwe-Saint-Pierre          Non    WOLUWE-SAINT-PIERRE   
16         1160              Auderghem          Non              AUDERGHEM   
17         1170    Watermael-Boitsfort          Non    WATERMAEL-BOITSFORT   
18         1180                  Uccle          Non                  UCCLE   
19         1190                 Forest          Non                 FOREST   
20         1200   Woluwe-Saint-Lambert          Non   WOLUWE-SAINT-LAMBERT   
21         1210  Saint-Josse-Ten-Noode          Non  SAINT-JOSSE-TEN-NOODE   

   Rôle linguistique  
0            NF - FN  
1            NF - FN  
2            NF - FN  
3            NF - FN  
4            NF - FN  
5            NF - FN  
6            NF - FN  
7            NF - FN  
8            NF - FN  
9            NF - FN  
10           NF - FN  
11           NF - FN  
12           NF - FN  
13           NF - FN  
14           NF - FN  
15           NF - FN  
16           NF - FN  
17           NF - FN  
18           NF - FN  
19           NF - FN  
20           NF - FN  
21           NF - FN

We'll keep the first two columns of the data frame

In [41]:
brussels_codes_df.drop(columns=['Sous-commune','Commune principale','Rôle linguistique'], inplace=True)
brussels_codes_df.head()

Code postal    Localité
0         1000   Bruxelles
1         1020      Laeken
2         1030  Schaerbeek
3         1040   Etterbeek
4         1050     Ixelles

I'll rename the columns to 'Postal Code' and 'Neighborhood'

In [42]:
brussels_codes_df.rename(columns={"Code postal": "Postal Code", "Localité":"Neighborhood"}, inplace=True)
brussels_codes_df.head()

Postal Code Neighborhood
0         1000    Bruxelles
1         1020       Laeken
2         1030   Schaerbeek
3         1040    Etterbeek
4         1050      Ixelles

In [43]:
print('There are {} neighborhoods in Brussels'.format(brussels_codes_df.shape[0]))

There are 22 neighborhoods in Brussels


We need the coordinates of each neighborhood in order to use the Foursquare API

### Paris neighborhoods coordinates

In [9]:

body = client_75203cb60a4e41d4bb1debcd2578d37c.get_object(Bucket='machinelearning-donotdelete-pr-7gneoytsamib8b',Key='paris_coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

paris_coordinates_df = pd.read_csv(body,';')
paris_coordinates_df


Postal Code   Latitude  Longitude
0         75001  48.860074   2.341092
1         75002  48.866729   2.340470
2         75003  48.863873   2.361786
3         75004  48.856053   2.356322
4         75005  48.846198   2.344591
5         75006  48.850714   2.332658
6         75007  48.856860   2.320032
7         75008  48.877364   2.317352
8         75009  48.872598   2.340453
9         75010  48.871792   2.357372
10        75011  48.857713   2.380543
11        75012  48.841114   2.388338
12        75013  48.832314   2.355485
13        75014  48.832866   2.326794
14        75015  48.841753   2.299553
15        75016  48.863864   2.276589
16        75017  48.884331   2.322106
17        75018  48.892618   2.345112
18        75019  48.882586   2.382156
19        75020  48.865088   2.398986

### Brussels neighborhoods coordinates

In [11]:

body = client_75203cb60a4e41d4bb1debcd2578d37c.get_object(Bucket='machinelearning-donotdelete-pr-7gneoytsamib8b',Key='brussels_coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

brussels_coordinates_df = pd.read_csv(body,';')
brussels_coordinates_df


Postal Code   Latitude  Longitude
0          1000  50.846714   4.352514
1          1020  50.876340   4.347956
2          1030  50.868056   4.373598
3          1040  50.834770   4.387934
4          1050  50.832551   4.367482
5          1060  50.824793   4.346348
6          1070  50.839148   4.328931
7          1080  50.854850   4.340254
8          1081  50.860935   4.331102
9          1082  50.864812   4.294890
10         1083  50.869365   4.308765
11         1090  50.875802   4.324237
12         1120  50.897867   4.385816
13         1130  50.892091   4.418116
14         1140  50.872615   4.402962
15         1150  50.837378   4.427669
16         1160  50.817025   4.427771
17         1170  50.798056   4.417055
18         1180  50.803547   4.333215
19         1190  50.811284   4.319380
20         1200  50.847449   4.435074
21         1210  50.850773   4.369453

For each city I'll merge the data frames containing neighborhoods names and coordinates.

In [19]:
paris_df = paris_codes_df.merge(paris_coordinates_df, left_on='Postal Code', right_on='Postal Code')
paris_df

Postal Code Neighborhood   Latitude  Longitude
0         75001      Paris 1  48.860074   2.341092
1         75002      Paris 2  48.866729   2.340470
2         75003      Paris 3  48.863873   2.361786
3         75004      Paris 4  48.856053   2.356322
4         75005      Paris 5  48.846198   2.344591
5         75006      Paris 6  48.850714   2.332658
6         75007      Paris 7  48.856860   2.320032
7         75008      Paris 8  48.877364   2.317352
8         75009      Paris 9  48.872598   2.340453
9         75010     Paris 10  48.871792   2.357372
10        75011     Paris 11  48.857713   2.380543
11        75012     Paris 12  48.841114   2.388338
12        75013     Paris 13  48.832314   2.355485
13        75014     Paris 14  48.832866   2.326794
14        75015     Paris 15  48.841753   2.299553
15        75016     Paris 16  48.863864   2.276589
16        75017     Paris 17  48.884331   2.322106
17        75018     Paris 18  48.892618   2.345112
18        75019     Paris 19  48.882586   2.382156
19        75020     Paris 20  48.865088   2.398986

In [44]:
brussels_df = brussels_codes_df.merge(brussels_coordinates_df, left_on='Postal Code', right_on='Postal Code')
brussels_df

Postal Code           Neighborhood   Latitude  Longitude
0          1000              Bruxelles  50.846714   4.352514
1          1020                 Laeken  50.876340   4.347956
2          1030             Schaerbeek  50.868056   4.373598
3          1040              Etterbeek  50.834770   4.387934
4          1050                Ixelles  50.832551   4.367482
5          1060           Saint-Gilles  50.824793   4.346348
6          1070             Anderlecht  50.839148   4.328931
7          1080   Molenbeek-Saint-Jean  50.854850   4.340254
8          1081             Koekelberg  50.860935   4.331102
9          1082  Berchem-Sainte-Agathe  50.864812   4.294890
10         1083              Ganshoren  50.869365   4.308765
11         1090                  Jette  50.875802   4.324237
12         1120    Neder-Over-Heembeek  50.897867   4.385816
13         1130                  Haren  50.892091   4.418116
14         1140                  Evere  50.872615   4.402962
15         1150    Woluwe-Saint-Pierre  50.837378   4.427669
16         1160              Auderghem  50.817025   4.427771
17         1170    Watermael-Boitsfort  50.798056   4.417055
18         1180                  Uccle  50.803547   4.333215
19         1190                 Forest  50.811284   4.319380
20         1200   Woluwe-Saint-Lambert  50.847449   4.435074
21         1210  Saint-Josse-Ten-Noode  50.850773   4.369453

### Foursquare API

We'll use Foursquare API in order to characterise each neighborhood of both cities

In [23]:
!pip -q install folium
import folium # map rendering library
print('folium imported')

folium imported


In [24]:
!pip install geopy
from geopy.geocoders import Nominatim
print('Nominatum imported')

Nominatum imported


In [25]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [31]:
address_p = 'Paris, FR'

geolocator_p = Nominatim(user_agent="paris_explorer")
location_p = geolocator_p.geocode(address_p)
latitude_p = location_p.latitude
longitude_p = location_p.longitude
print('The geograpical coordinate of Paris are {}, {}.'.format(latitude_p, longitude_p))

The geograpical coordinate of Paris are 48.8566969, 2.3514616.


In [32]:
address_b = 'Brussels, BE'

geolocator_b = Nominatim(user_agent="brussels_explorer")
location_b = geolocator_b.geocode(address_b)
latitude_b = location_b.latitude
longitude_b = location_b.longitude
print('The geograpical coordinate of Brussels are {}, {}.'.format(latitude_b, longitude_b))

The geograpical coordinate of Brussels are 50.8465573, 4.351697.


In [48]:
# create map of Paris using latitude and longitude values
map_of_paris = folium.Map(location=[latitude_p, longitude_p], zoom_start=12)

# add markers to map
for lat, lng, neighborhood in zip(paris_df['Latitude'], paris_df['Longitude'], paris_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_of_paris)  
    
map_of_paris

In [50]:
# create map of Brussels using latitude and longitude values
map_of_brussels = folium.Map(location=[latitude_b, longitude_b], zoom_start=12)

# add markers to map
for lat, lng, neighborhood in zip(brussels_df['Latitude'], brussels_df['Longitude'], brussels_df['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_of_brussels)  
    
map_of_brussels

I'll now use Foursquare to get the venues of each neighborhood for both cities.

In [78]:
# The code was removed by Watson Studio for sharing.

In [52]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

In [53]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [54]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        if (len(requests.get(url).json()["response"])== 0):
            results = {}
        else:
            results = requests.get(url).json()["response"]['groups'][0]['items']
        #results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [55]:
paris_venues = getNearbyVenues(names=paris_df['Neighborhood'],
                                  latitudes=paris_df['Latitude'],
                                  longitudes=paris_df['Longitude'])

Paris 1
Paris 2
Paris 3
Paris 4
Paris 5
Paris 6
Paris 7
Paris 8
Paris 9
Paris 10
Paris 11
Paris 12
Paris 13
Paris 14
Paris 15
Paris 16
Paris 17
Paris 18
Paris 19
Paris 20


In [56]:
print(paris_venues.shape)
paris_venues.head()

(1349, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      Paris 1              48.860074                2.341092   
1      Paris 1              48.860074                2.341092   
2      Paris 1              48.860074                2.341092   
3      Paris 1              48.860074                2.341092   
4      Paris 1              48.860074                2.341092   

                                               Venue  Venue Latitude  \
0                              Cour Carrée du Louvre       48.860360   
1                                    Place du Louvre       48.859841   
2  Église Saint-Germain-l'Auxerrois (Église Saint...       48.859520   
3                                      Coffee Crêpes       48.858841   
4                                  Boutique yam'Tcha       48.861710   

   Venue Longitude      Venue Category  
0         2.338543    Pedestrian Plaza  
1         2.340822               Plaza  
2         2.341306              Church  
3         2.340802         Coffee Shop  
4         2.342380  Chinese Restaurant

In [71]:
brussels_venues = getNearbyVenues(names=brussels_df['Neighborhood'],
                                  latitudes=brussels_df['Latitude'],
                                  longitudes=brussels_df['Longitude'])

Bruxelles
Laeken
Schaerbeek
Etterbeek
Ixelles
Saint-Gilles
Anderlecht
Molenbeek-Saint-Jean
Koekelberg
Berchem-Sainte-Agathe
Ganshoren
Jette
Neder-Over-Heembeek
Haren
Evere
Woluwe-Saint-Pierre
Auderghem
Watermael-Boitsfort
Uccle
Forest
Woluwe-Saint-Lambert
Saint-Josse-Ten-Noode


In [72]:
print(brussels_venues.shape)
brussels_venues.head()

(736, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Bruxelles              50.846714                4.352514   
1    Bruxelles              50.846714                4.352514   
2    Bruxelles              50.846714                4.352514   
3    Bruxelles              50.846714                4.352514   
4    Bruxelles              50.846714                4.352514   

                                               Venue  Venue Latitude  \
0            Grand Place / Grote Markt (Grote Markt)       50.846776   
1                               AB Ancienne Belgique       50.847143   
2  Galerie de la Reine / Koninginnegalerij (Konin...       50.847297   
3                            Aux Merveilleux de Fred       50.848117   
4                                  Mary Chocolaterie       50.847541   

   Venue Longitude  Venue Category  
0         4.352481           Plaza  
1         4.349485    Concert Hall  
2         4.354601   Shopping Mall  
3         4.352141    Dessert Shop  
4         4.354785  Chocolate Shop

In [73]:
paris_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                         
Paris 1                          60                      60     60   
Paris 10                         57                      57     57   
Paris 11                         85                      85     85   
Paris 12                         41                      41     41   
Paris 13                         53                      53     53   
Paris 14                         46                      46     46   
Paris 15                         44                      44     44   
Paris 16                         35                      35     35   
Paris 17                         99                      99     99   
Paris 18                         56                      56     56   
Paris 19                         30                      30     30   
Paris 2                         100                     100    100   
Paris 20                         55                      55     55   
Paris 3                          95                      95     95   
Paris 4                         100                     100    100   
Paris 5                         100                     100    100   
Paris 6                          61                      61     61   
Paris 7                          77                      77     77   
Paris 8                         100                     100    100   
Paris 9                          55                      55     55   

              Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                   
Paris 1                   60               60              60  
Paris 10                  57               57              57  
Paris 11                  85               85              85  
Paris 12                  41               41              41  
Paris 13                  53               53              53  
Paris 14                  46               46              46  
Paris 15                  44               44              44  
Paris 16                  35               35              35  
Paris 17                  99               99              99  
Paris 18                  56               56              56  
Paris 19                  30               30              30  
Paris 2                  100              100             100  
Paris 20                  55               55              55  
Paris 3                   95               95              95  
Paris 4                  100              100             100  
Paris 5                  100              100             100  
Paris 6                   61               61              61  
Paris 7                   77               77              77  
Paris 8                  100              100             100  
Paris 9                   55               55              55

In [76]:
print('There are {} uniques categories in Paris.'.format(len(paris_venues['Venue Category'].unique())))

There are 202 uniques categories in Paris.


In [74]:
brussels_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                  
Anderlecht                                23                      23     23   
Auderghem                                 43                      43     43   
Berchem-Sainte-Agathe                     16                      16     16   
Bruxelles                                100                     100    100   
Etterbeek                                 40                      40     40   
Evere                                     18                      18     18   
Forest                                    15                      15     15   
Ganshoren                                 29                      29     29   
Haren                                     10                      10     10   
Ixelles                                   46                      46     46   
Jette                                     48                      48     48   
Koekelberg                                21                      21     21   
Laeken                                    20                      20     20   
Molenbeek-Saint-Jean                      36                      36     36   
Neder-Over-Heembeek                       13                      13     13   
Saint-Gilles                              38                      38     38   
Saint-Josse-Ten-Noode                     56                      56     56   
Schaerbeek                                36                      36     36   
Uccle                                     34                      34     34   
Watermael-Boitsfort                       35                      35     35   
Woluwe-Saint-Lambert                      41                      41     41   
Woluwe-Saint-Pierre                       18                      18     18   

                       Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                            
Anderlecht                         23               23              23  
Auderghem                          43               43              43  
Berchem-Sainte-Agathe              16               16              16  
Bruxelles                         100              100             100  
Etterbeek                          40               40              40  
Evere                              18               18              18  
Forest                             15               15              15  
Ganshoren                          29               29              29  
Haren                              10               10              10  
Ixelles                            46               46              46  
Jette                              48               48              48  
Koekelberg                         21               21              21  
Laeken                             20               20              20  
Molenbeek-Saint-Jean               36               36              36  
Neder-Over-Heembeek                13               13              13  
Saint-Gilles                       38               38              38  
Saint-Josse-Ten-Noode              56               56              56  
Schaerbeek                         36               36              36  
Uccle                              34               34              34  
Watermael-Boitsfort                35               35              35  
Woluwe-Saint-Lambert               41               41              41  
Woluwe-Saint-Pierre                18               18              18

In [77]:
print('There are {} uniques categories in Brussels.'.format(len(brussels_venues['Venue Category'].unique())))

There are 181 uniques categories in Brussels.


In the next part of the project I'll merge all the neighborhoods and segment them in clusters trying to find similar neghborhoods independently of the city they belong to.